File: back-end/ShopHangTet/Services/ProductService.cs
using Microsoft.EntityFrameworkCore;
using ShopHangTet.Data;
using ShopHangTet.DTOs;
using ShopHangTet.Models;

namespace ShopHangTet.Services
{
    public class ProductService : IProductService
    {
        private readonly ShopHangTetDbContext _context;

        public ProductService(ShopHangTetDbContext context)
        {
            _context = context;
        }

        private static int CalculateGiftBoxAvailableStock(GiftBox giftBox, Dictionary<string, Item> itemMap)
        {
            if (giftBox.Items == null || giftBox.Items.Count == 0)
            {
                return 0;
            }

            var minAvailable = int.MaxValue;

            foreach (var giftItem in giftBox.Items)
            {
                if (giftItem.Quantity <= 0)
                {
                    continue;
                }

                if (!itemMap.TryGetValue(giftItem.ItemId, out var item))
                {
                    return 0;
                }

                var available = Math.Max(0, item.AvailableQuantity / giftItem.Quantity);
                minAvailable = Math.Min(minAvailable, available);
            }

            return minAvailable == int.MaxValue ? 0 : minAvailable;
        }

        public async Task<List<Item>> GetItemsAsync(string? name = null)
        {
            var query = _context.Items.Where(x => x.IsActive);
            if (!string.IsNullOrWhiteSpace(name))
            {
                // Normalize search keyword.
                var normalizedName = name.Trim().ToLower();

                query = query.Where(x => x.Name.ToLower().Contains(normalizedName));
            }

            var items = await query.ToListAsync();
            return items.Select(x => new Item
            {
                Id = x.Id,
                Name = x.Name,
                Category = x.Category,
                Price = x.Price,
                Images = x.Images,
                IsAlcohol = x.IsAlcohol,
                StockQuantity = Math.Max(0, x.AvailableQuantity),
                ReservedQuantity = x.ReservedQuantity,
                IsActive = x.IsActive,
                CreatedAt = x.CreatedAt
            }).ToList();
        }

        public async Task<Item?> GetItemByIdAsync(string id)
        {
            return await _context.Items
                .FirstOrDefaultAsync(x => x.Id == id && x.IsActive);
        }

        public async Task<List<GiftBoxListDto>> GetGiftBoxesAsync(string? name = null)
        {
            var query = _context.GiftBoxes.Where(x => x.IsActive);

            if (!string.IsNullOrWhiteSpace(name))
            {
                // Normalize search keyword.
                var normalizedName = name.Trim().ToLower();
                query = query.Where(x => x.Name.ToLower().Contains(normalizedName));
            }

            var giftBoxes = await query.ToListAsync();
            var collectionIds = giftBoxes.Select(x => x.CollectionId).Distinct().ToList();
            var collections = await _context.Collections
                .Where(x => collectionIds.Contains(x.Id) && x.IsActive)
                .ToDictionaryAsync(x => x.Id, x => x.Name);

            var itemIds = giftBoxes
                .SelectMany(x => x.Items)
                .Select(i => i.ItemId)
                .Distinct()
                .ToList();
            var items = await _context.Items
                .Where(x => itemIds.Contains(x.Id) && x.IsActive)
                .ToDictionaryAsync(x => x.Id, x => x);

            var tagIds = giftBoxes
                .SelectMany(x => x.Tags)
                .Distinct()
                .ToList();
            var tags = await _context.Tags
                .Where(x => tagIds.Contains(x.Id) && x.IsActive)
                .ToDictionaryAsync(x => x.Id, x => x.Name);

            return giftBoxes
                .Select(x => new GiftBoxListDto
                {
                    Id = x.Id,
                    Name = x.Name,
                    Description = x.Description,
                    Price = x.Price,
                    StockQuantity = CalculateGiftBoxAvailableStock(x, items),
                    Image = x.Images.FirstOrDefault(),
                    Collection = collections.TryGetValue(x.CollectionId, out var collectionName) ? collectionName : string.Empty,
                    Tags = x.Tags
                        .Select(tagId => tags.TryGetValue(tagId, out var tagName) ? tagName : string.Empty)
                        .Where(tagName => !string.IsNullOrWhiteSpace(tagName))
                        .ToList(),
                    Items = x.Items.Select(gi =>
                    {
                        items.TryGetValue(gi.ItemId, out var item);
                        return new GiftBoxItemFlatDto
                        {
                            Id = gi.ItemId,
                            Name = item?.Name ?? string.Empty,
                            Price = gi.ItemPriceSnapshot > 0 ? gi.ItemPriceSnapshot : item?.Price ?? 0,
                            Image = item?.Images.FirstOrDefault(),
                            Quantity = gi.Quantity
                        };
                    }).ToList(),
                    IsActive = x.IsActive,
                    CreatedAt = x.CreatedAt
                })
                .ToList();
        }

        public async Task<GiftBoxDetailDto?> GetGiftBoxDetailByIdAsync(string id)
        {
            var giftBox = await _context.GiftBoxes
                .FirstOrDefaultAsync(x => x.Id == id && x.IsActive);

            if (giftBox == null)
            {
                return null;
            }

            var itemIds = giftBox.Items.Select(i => i.ItemId).Distinct().ToList();
            var items = await _context.Items
                .Where(it => itemIds.Contains(it.Id) && it.IsActive)
                .ToDictionaryAsync(it => it.Id, it => it);

            var collection = await _context.Collections
                .FirstOrDefaultAsync(x => x.Id == giftBox.CollectionId && x.IsActive);

            var tagIds = giftBox.Tags.Distinct().ToList();
            var tags = await _context.Tags
                .Where(t => tagIds.Contains(t.Id) && t.IsActive)
                .ToDictionaryAsync(t => t.Id, t => t.Name);

            return new GiftBoxDetailDto
            {
                Id = giftBox.Id,
                Name = giftBox.Name,
                Description = giftBox.Description,
                Price = giftBox.Price,
                StockQuantity = CalculateGiftBoxAvailableStock(giftBox, items),
                Image = giftBox.Images.FirstOrDefault(),
                Images = giftBox.Images,
                Collection = collection?.Name ?? string.Empty,
                Tags = giftBox.Tags
                    .Select(tagId => tags.TryGetValue(tagId, out var tagName) ? tagName : string.Empty)
                    .Where(tagName => !string.IsNullOrWhiteSpace(tagName))
                    .ToList(),
                Items = giftBox.Items.Select(gi =>
                {
                    items.TryGetValue(gi.ItemId, out var item);
                    return new GiftBoxItemFlatDto
                    {
                        Id = gi.ItemId,
                        Name = item?.Name ?? string.Empty,
                        Price = gi.ItemPriceSnapshot > 0 ? gi.ItemPriceSnapshot : item?.Price ?? 0,
                        Image = item?.Images.FirstOrDefault(),
                        Quantity = gi.Quantity
                    };
                }).ToList(),
                IsActive = giftBox.IsActive,
                CreatedAt = giftBox.CreatedAt
            };
        }

        public async Task<List<CollectionListDto>> GetCollectionsAsync(string? name = null)
        {
            var query = _context.Collections.Where(x => x.IsActive);

            if (!string.IsNullOrWhiteSpace(name))
            {
                // Normalize search keyword.
                var normalizedName = name.Trim().ToLower();

                query = query.Where(x => x.Name.ToLower().Contains(normalizedName));
            }

            var collections = await query
                .OrderBy(x => x.DisplayOrder)
                .ToListAsync();

            var collectionIds = collections.Select(x => x.Id).ToList();
            var giftBoxes = await _context.GiftBoxes
                .Where(x => x.IsActive && collectionIds.Contains(x.CollectionId))
                .ToListAsync();

            return collections.Select(collection => new CollectionListDto
            {
                Id = collection.Id,
                Name = collection.Name,
                Description = collection.Description,
                CoverImage = collection.CoverImage,
                PricingMultiplier = collection.PricingMultiplier,
                PackagingFee = collection.PackagingFee,
                DisplayOrder = collection.DisplayOrder,
                IsActive = collection.IsActive,
                CreatedAt = collection.CreatedAt,
                GiftBoxCount = giftBoxes.Count(x => x.CollectionId == collection.Id)
            }).ToList();
        }

        public async Task<CollectionDetailDto?> GetCollectionDetailByIdAsync(string id)
        {
            var collection = await _context.Collections
                .FirstOrDefaultAsync(x => x.Id == id && x.IsActive);

            if (collection == null)
            {
                return null;
            }

            var giftBoxes = await _context.GiftBoxes
                .Where(x => x.CollectionId == id && x.IsActive)
                .Select(x => new GiftBoxListDto
                {
                    Id = x.Id,
                    Name = x.Name,
                    Description = x.Description,
                    Price = x.Price,
                    Image = x.Images.FirstOrDefault(),
                    Collection = collection.Name,
                    IsActive = x.IsActive,
                    CreatedAt = x.CreatedAt,
                    Tags = new List<string>(),
                    Items = new List<GiftBoxItemFlatDto>()
                })
                .ToListAsync();

            return new CollectionDetailDto
            {
                Id = collection.Id,
                Name = collection.Name,
                Description = collection.Description,
                CoverImage = collection.CoverImage,
                PricingMultiplier = collection.PricingMultiplier,
                PackagingFee = collection.PackagingFee,
                DisplayOrder = collection.DisplayOrder,
                IsActive = collection.IsActive,
                CreatedAt = collection.CreatedAt,
                GiftBoxes = giftBoxes
            };
        }

        // GiftBox pricing

        /// Price = (item cost * multiplier) + packaging fee.
        public async Task<decimal> CalculateGiftBoxPriceAsync(string collectionId, List<GiftBoxItem> items)
        {
            var collection = await _context.Collections.FirstOrDefaultAsync(x => x.Id == collectionId);
            if (collection == null)
                throw new InvalidOperationException("Collection not found");

            var itemIds = items.Select(x => x.ItemId).Distinct().ToList();
            var dbItems = await _context.Items.Where(x => itemIds.Contains(x.Id)).ToListAsync();
            var itemMap = dbItems.ToDictionary(x => x.Id, x => x);

            decimal totalCost = 0;
            foreach (var giftBoxItem in items)
            {
                if (itemMap.TryGetValue(giftBoxItem.ItemId, out var item))
                {
                    totalCost += item.Price * giftBoxItem.Quantity;
                }
            }

            // Apply collection pricing rule.
            var price = (totalCost * collection.PricingMultiplier) + collection.PackagingFee;
            return Math.Round(price, 0); // Round to VND.
        }

        /// Create GiftBox with calculated price by default.
        public async Task<GiftBox> CreateGiftBoxAsync(CreateGiftBoxDto dto)
        {
            decimal price;
            var itemIds = dto.Items.Select(i => i.ItemId).Distinct().ToList();
            var dbItems = await _context.Items.Where(x => itemIds.Contains(x.Id)).ToListAsync();
            var itemMap = dbItems.ToDictionary(x => x.Id, x => x);

            var giftBoxItems = dto.Items.Select(i => new GiftBoxItem
            {
                ItemId = i.ItemId,
                Quantity = i.Quantity,
                ItemPriceSnapshot = itemMap.TryGetValue(i.ItemId, out var item) ? item.Price : 0
            }).ToList();

            if (dto.PriceOverride.HasValue && dto.PriceOverride.Value > 0)
            {
                price = dto.PriceOverride.Value;
            }
            else
            {
                price = await CalculateGiftBoxPriceAsync(dto.CollectionId, giftBoxItems);
            }

            var giftBox = new GiftBox
            {
                Name = dto.Name,
                Description = dto.Description,
                Price = price,
                Images = dto.Images,
                CollectionId = dto.CollectionId,
                Tags = dto.Tags,
                Items = giftBoxItems,
                IsActive = true,
                CreatedAt = DateTime.UtcNow
            };

            _context.GiftBoxes.Add(giftBox);
            await _context.SaveChangesAsync();

            return giftBox;
        }

        /// Recalculate price when items or collection change.
        public async Task<GiftBox> UpdateGiftBoxAsync(string id, UpdateGiftBoxDto dto)
        {
            var giftBox = await _context.GiftBoxes.FirstOrDefaultAsync(x => x.Id == id);
            if (giftBox == null) throw new InvalidOperationException("GiftBox not found");

            if (!string.IsNullOrEmpty(dto.Name)) giftBox.Name = dto.Name;
            if (!string.IsNullOrEmpty(dto.Description)) giftBox.Description = dto.Description;
            if (dto.Images != null) giftBox.Images = dto.Images;
            if (!string.IsNullOrEmpty(dto.CollectionId)) giftBox.CollectionId = dto.CollectionId;
            if (dto.Tags != null) giftBox.Tags = dto.Tags;
            if (dto.IsActive.HasValue) giftBox.IsActive = dto.IsActive.Value;

            if (dto.Items != null)
            {
                var itemIds = dto.Items.Select(i => i.ItemId).Distinct().ToList();
                var dbItems = await _context.Items.Where(x => itemIds.Contains(x.Id)).ToListAsync();
                var itemMap = dbItems.ToDictionary(x => x.Id, x => x);

                giftBox.Items = dto.Items.Select(i => new GiftBoxItem
                {
                    ItemId = i.ItemId,
                    Quantity = i.Quantity,
                    ItemPriceSnapshot = itemMap.TryGetValue(i.ItemId, out var item) ? item.Price : 0
                }).ToList();
            }

            // Recalculate price when needed.
            if (dto.PriceOverride.HasValue && dto.PriceOverride.Value > 0)
            {
                giftBox.Price = dto.PriceOverride.Value;
            }
            else if (dto.Items != null || dto.CollectionId != null)
            {
                giftBox.Price = await CalculateGiftBoxPriceAsync(giftBox.CollectionId, giftBox.Items);
            }

            await _context.SaveChangesAsync();
            return giftBox;
        }
    }
}

File: back-end-nodejs/src/services/productService.js
import { Item, GiftBox, Collection, Tag } from '../models/index.js';

/**
 * Product Service - Tương đương ProductService.cs
 */
export class ProductService {
    // ========== Items ==========

    async getItems(name = null) {
        const query = { isActive: true };
        if (name && name.trim()) {
            query.name = { $regex: name.trim(), $options: 'i' };
        }
        return Item.find(query);
    }

    async getItemById(id) {
        return Item.findOne({ _id: id, isActive: true });
    }

    // ========== Gift Boxes ==========

    async getGiftBoxes(name = null) {
        const query = { isActive: true };
        if (name && name.trim()) {
            query.name = { $regex: name.trim(), $options: 'i' };
        }

        const giftBoxes = await GiftBox.find(query);

        // Batch load collections
        const collectionIds = [...new Set(giftBoxes.map((x) => x.collectionId))];
        const collections = await Collection.find({ _id: { $in: collectionIds }, isActive: true });
        const collectionMap = {};
        collections.forEach((c) => (collectionMap[c._id.toString()] = c.name));

        // Batch load items
        const itemIds = [...new Set(giftBoxes.flatMap((x) => x.items.map((i) => i.itemId)))];
        const items = await Item.find({ _id: { $in: itemIds }, isActive: true });
        const itemMap = {};
        items.forEach((i) => (itemMap[i._id.toString()] = i));

        // Batch load tags
        const tagIds = [...new Set(giftBoxes.flatMap((x) => x.tags))];
        const tags = await Tag.find({ _id: { $in: tagIds }, isActive: true });
        const tagMap = {};
        tags.forEach((t) => (tagMap[t._id.toString()] = t.name));

        return giftBoxes.map((x) => ({
            Id: x._id.toString(),
            Name: x.name,
            Description: x.description,
            Price: x.price,
            Image: x.images?.[0] || null,
            Collection: collectionMap[x.collectionId] || '',
            Tags: x.tags
                .map((tagId) => tagMap[tagId] || '')
                .filter((t) => t),
            Items: x.items.map((gi) => {
                const item = itemMap[gi.itemId];
                return {
                    Id: gi.itemId,
                    Name: item?.name || '',
                    Price: gi.itemPriceSnapshot > 0 ? gi.itemPriceSnapshot : item?.price || 0,
                    Image: item?.images?.[0] || null,
                    Quantity: gi.quantity,
                };
            }),
            IsActive: x.isActive,
            CreatedAt: x.createdAt,
        }));
    }

    async getGiftBoxDetailById(id) {
        const giftBox = await GiftBox.findOne({ _id: id, isActive: true });
        if (!giftBox) return null;

        // Load item details
        const itemIds = giftBox.items.map((i) => i.itemId);
        const items = await Item.find({ _id: { $in: itemIds }, isActive: true });
        const itemMap = {};
        items.forEach((i) => (itemMap[i._id.toString()] = i));

        // Load collection
        const collection = await Collection.findOne({ _id: giftBox.collectionId, isActive: true });

        // Load tags
        const tagIds = [...new Set(giftBox.tags)];
        const tags = await Tag.find({ _id: { $in: tagIds }, isActive: true });
        const tagMap = {};
        tags.forEach((t) => (tagMap[t._id.toString()] = t.name));

        return {
            Id: giftBox._id.toString(),
            Name: giftBox.name,
            Description: giftBox.description,
            Price: giftBox.price,
            Image: giftBox.images?.[0] || null,
            Images: giftBox.images || [],
            Collection: collection?.name || '',
            Tags: giftBox.tags
                .map((tagId) => tagMap[tagId] || '')
                .filter((t) => t),
            Items: giftBox.items.map((gi) => {
                const item = itemMap[gi.itemId];
                return {
                    Id: gi.itemId,
                    Name: item?.name || '',
                    Price: gi.itemPriceSnapshot > 0 ? gi.itemPriceSnapshot : item?.price || 0,
                    Image: item?.images?.[0] || null,
                    Quantity: gi.quantity,
                };
            }),
            IsActive: giftBox.isActive,
            CreatedAt: giftBox.createdAt,
        };
    }

    // ========== Collections ==========

    async getCollections(name = null) {
        const query = { isActive: true };
        if (name && name.trim()) {
            query.name = { $regex: name.trim(), $options: 'i' };
        }

        const collections = await Collection.find(query).sort({ displayOrder: 1 });

        const collectionIds = collections.map((c) => c._id.toString());
        const giftBoxes = await GiftBox.find({
            isActive: true,
            collectionId: { $in: collectionIds },
        });

        return collections.map((c) => ({
            Id: c._id.toString(),
            Name: c.name,
            Description: c.description,
            CoverImage: c.coverImage,
            PricingMultiplier: c.pricingMultiplier,
            PackagingFee: c.packagingFee,
            DisplayOrder: c.displayOrder,
            IsActive: c.isActive,
            CreatedAt: c.createdAt,
            GiftBoxCount: giftBoxes.filter((gb) => gb.collectionId === c._id.toString()).length,
        }));
    }

    async getCollectionDetailById(id) {
        const collection = await Collection.findOne({ _id: id, isActive: true });
        if (!collection) return null;

        const giftBoxes = await GiftBox.find({ collectionId: id, isActive: true });

        return {
            Id: collection._id.toString(),
            Name: collection.name,
            Description: collection.description,
            CoverImage: collection.coverImage,
            PricingMultiplier: collection.pricingMultiplier,
            PackagingFee: collection.packagingFee,
            DisplayOrder: collection.displayOrder,
            IsActive: collection.isActive,
            CreatedAt: collection.createdAt,
            GiftBoxes: giftBoxes.map((x) => ({
                Id: x._id.toString(),
                Name: x.name,
                Description: x.description,
                Price: x.price,
                Image: x.images?.[0] || null,
                Collection: collection.name,
                Tags: [],
                Items: [],
                IsActive: x.isActive,
                CreatedAt: x.createdAt,
            })),
        };
    }

    // ========== GiftBox Pricing ==========

    /**
     * Tính giá GiftBox tự động: (sum item cost × multiplier) + packagingFee
     */
    async calculateGiftBoxPrice(collectionId, items) {
        const collection = await Collection.findById(collectionId);
        if (!collection) throw new Error('Collection not found');

        const itemIds = [...new Set(items.map((x) => x.ItemId || x.itemId))];
        const dbItems = await Item.find({ _id: { $in: itemIds } });
        const itemMap = {};
        dbItems.forEach((i) => (itemMap[i._id.toString()] = i));

        let totalCost = 0;
        for (const giftBoxItem of items) {
            const itemId = giftBoxItem.ItemId || giftBoxItem.itemId;
            const item = itemMap[itemId];
            if (item) {
                totalCost += item.price * (giftBoxItem.Quantity || giftBoxItem.quantity || 0);
            }
        }

        const price = totalCost * collection.pricingMultiplier + collection.packagingFee;
        return Math.round(price);
    }

    // ========== GiftBox CRUD (ADMIN) ==========

    async createGiftBox(dto) {
        const itemIds = [...new Set(dto.Items.map((i) => i.ItemId))];
        const dbItems = await Item.find({ _id: { $in: itemIds } });
        const itemMap = {};
        dbItems.forEach((i) => (itemMap[i._id.toString()] = i));

        const giftBoxItems = dto.Items.map((i) => ({
            itemId: i.ItemId,
            quantity: i.Quantity,
            itemPriceSnapshot: itemMap[i.ItemId]?.price || 0,
        }));

        let price;
        if (dto.PriceOverride && dto.PriceOverride > 0) {
            price = dto.PriceOverride;
        } else {
            price = await this.calculateGiftBoxPrice(dto.CollectionId, giftBoxItems);
        }

        const giftBox = await GiftBox.create({
            name: dto.Name,
            description: dto.Description || '',
            price,
            images: dto.Images || [],
            collectionId: dto.CollectionId,
            tags: dto.Tags || [],
            items: giftBoxItems,
            isActive: true,
        });

        return giftBox;
    }

    async updateGiftBox(id, dto) {
        const giftBox = await GiftBox.findById(id);
        if (!giftBox) throw new Error('GiftBox not found');

        if (dto.Name) giftBox.name = dto.Name;
        if (dto.Description) giftBox.description = dto.Description;
        if (dto.Images) giftBox.images = dto.Images;
        if (dto.CollectionId) giftBox.collectionId = dto.CollectionId;
        if (dto.Tags) giftBox.tags = dto.Tags;
        if (dto.IsActive !== undefined && dto.IsActive !== null) giftBox.isActive = dto.IsActive;

        if (dto.Items) {
            const itemIds = [...new Set(dto.Items.map((i) => i.ItemId))];
            const dbItems = await Item.find({ _id: { $in: itemIds } });
            const itemMap = {};
            dbItems.forEach((i) => (itemMap[i._id.toString()] = i));

            giftBox.items = dto.Items.map((i) => ({
                itemId: i.ItemId,
                quantity: i.Quantity,
                itemPriceSnapshot: itemMap[i.ItemId]?.price || 0,
            }));
        }

        // Recalculate price
        if (dto.PriceOverride && dto.PriceOverride > 0) {
            giftBox.price = dto.PriceOverride;
        } else if (dto.Items || dto.CollectionId) {
            giftBox.price = await this.calculateGiftBoxPrice(giftBox.collectionId, giftBox.items);
        }

        await giftBox.save();
        return giftBox;
    }
}
